### ISSUE TITLE
Backup listing whole tree even though subtree is excluded

### ISSUE THREAD

**Issue by SimplicityApks** [restic/restic#4418]
Output of `restic version`
--------------------------
restic 0.15.2 compiled with go1.20.3 on linux/amd64

How did you run restic exactly?
-------------------------------

On a large drive or fileserver with lots of files and deep tree structure:

    restic -r REPO backup --exclude-file=EXCLUDEFILE /volume1

with an EXCLUDEFILE containing subdirectories, such as 
```
**/tmp/**
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------

any backend

Expected behavior
-----------------

Restic does not even touch anything inside excluded directories.

Actual behavior
---------------

Haven't traced it, but from what I can see in the source code restic will scan the whole filesystem twice, once in the [scanner](https://github.com/restic/restic/blob/master/internal/archiver/scanner.go#L60) (which I know you can disable with `--no-scan`) and once for the actual backup in the [archiver](https://github.com/restic/restic/blob/master/internal/archiver/archiver.go#L612). 

On a system with very many files and I/O limits the scanning alone will take days (I've seen `scan finished in 85303.814s: 22030498 files, 22.134 TiB`), and I would assume the backup is also significantly slowed down.
It would be nice to be able to exclude large directory trees completely, so that they are not slowing down this process. 
I know about `--files-from` which would allow building that externally, but that has lots of other problems.

Steps to reproduce the behavior
-------------------------------

Run restic backup with excluded dirs and see that it will try to walk the excluded tree.

Do you have any idea what may have caused this?
-----------------------------------------------

The problems is excludes afaik are not considered for directories.

Do you have an idea how to solve the issue?
-------------------------------------------

I'd propose checking for excludes earlier in the scanning/archiving process. If the directory is excluded, it should not be walked.
The tricky part is negative excludes with exclamation marks. 
However, the [docs](https://restic.readthedocs.io/en/latest/040_backup.html) state:
> once a directory is excluded, it is not possible to include files inside the directory.

So I'd assume that is not an issue.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Love restic, saves so much money and time with deduplication & compression ;)

**Comment by MichaelEischer** [comment]
> Haven't traced it, but from what I can see in the source code restic will scan the whole filesystem twice, once in the [scanner](https://github.com/restic/restic/blob/master/internal/archiver/scanner.go#L60) (which I know you can disable with `--no-scan`) and once for the actual backup in the [archiver](https://github.com/restic/restic/blob/master/internal/archiver/archiver.go#L612).

As the operating system caches filesystem metadata, scanning the filesystem twice doesn't automatically require twice the time.

> **/tmp/**

That pattern is equivalent to `tmp`. Restic already checks excluded paths before opening any files/folders. (There's no optimization to shortcut `tmp/*`, but that isn't relevant for your example. And even in that case restic would only list all filenames in the `tmp` folder without actually accessing the files).

In other words, I can't reproduce the issue.

**Comment by SimplicityApks** [comment]
> As the operating system caches filesystem metadata, scanning the filesystem twice doesn't automatically require twice the time.

Probably true, but can't be relied on for my case of millions of files (metadata likely exceeding RAM). But not sure. This double checking is not the main problem (but could of course also be optimized).

> 
> > **/tmp/**
> 
> That pattern is equivalent to `tmp`. Restic already checks excluded paths before opening any files/folders. (There's no optimization to shortcut `tmp/*`, but that isn't relevant for your example. And even in that case restic would only list all filenames in the `tmp` folder without actually accessing the files).
> 
> In other words, I can't reproduce the issue.

yeah sure, it can be the same, just had the stars for clarity. Can also be `/tmp` or `/tmp/**`. The point is exactly your statement "even then restic would only list all filenames". Why is this needed? It can take forever on fileshares with many files (`node_modules` is horrible in that regard) and I regurlarly see restic backup going down from 100-300MB reads to only 1MB (probably when it hits those bottleneck folders with gigantic tree where everything is ignored).

I feel checking the exclude list for every folder before walking in twice is not too crazy of an optimization :D

**Comment by MichaelEischer** [comment]
> I feel checking the exclude list for every folder before walking in twice is not too crazy of an optimization :D

If you use an exclude pattern `node_modules` then restic completely skips that folder an never opens/enters it. Only in case of `node_modules/*` restic will have to list all files in the folder (this only requires reading the folder content, but won't access any item in that folder. That is much faster than running `ls` in such a folder. It's only a few `getdents64` calls compared to a `stat` call for each directory entry).

Skipping the iteration through the directory isn't that trivial: for a pattern `node_modules/*`, our pattern matcher would have to provide the information that everything in `node_modules` will be excluded. That is, it is not a matter of checking the excludes earlier on, but rather of deriving statements for all files in a folder based on a pattern. Although, implementing that probably isn't too hard, although it might run into quite a few corner cases.

**Comment by MichaelEischer** [comment]
> (probably when it hits those bottleneck folders with gigantic tree where everything is ignored).

Do you have any data to back up that claim? Just reading a bunch of small files would produce the exact same effect.

**Comment by SimplicityApks** [comment]
ok you are indeed right. I had some time to try around, and cannot reproduce restic listing files inside excluded folders. 
Something in the source code tripped me off in the few minutes i had digging through it.

So yes I guess it is running fine, and reading exactly what it is supposed to. (scanning twice you can easily disable with `--no-scan` so that is not a big issue for me).

Thanks for the argument ;)
